src/midi/midi_fluidsynth.cpp

```diff
@@ -519,122 +543,146 @@ void MidiHandlerFluidsynth::Close()
 	// Reset the members
 	synth.reset();
 	settings.reset();
-	last_played_frame = 0;
 	selected_font.clear();
 
+	last_rendered_ms = 0.0;
+	ms_per_frame     = 0.0;
+
 	is_open = false;
 }
 
-void MidiHandlerFluidsynth::PlayMsg(const uint8_t *msg)
+uint16_t MidiHandlerFluidsynth::GetNumPendingFrames()
 {
+	const auto now_ms = PIC_FullIndex();
+
+	// Wake up the channel and update the last rendered time datum.
 	assert(channel);
-	channel->WakeUp();
-
-	const int chanID = msg[0] & 0b1111;
-
-	switch (msg[0] & 0b1111'0000) {
-	case 0b1000'0000:
-		fluid_synth_noteoff(synth.get(), chanID, msg[1]);
-		break;
-	case 0b1001'0000:
-		fluid_synth_noteon(synth.get(), chanID, msg[1], msg[2]);
-		break;
-	case 0b1010'0000:
-		fluid_synth_key_pressure(synth.get(), chanID, msg[1], msg[2]);
-		break;
-	case 0b1011'0000:
-		fluid_synth_cc(synth.get(), chanID, msg[1], msg[2]);
-		break;
-	case 0b1100'0000:
-		fluid_synth_program_change(synth.get(), chanID, msg[1]);
-		break;
-	case 0b1101'0000:
-		fluid_synth_channel_pressure(synth.get(), chanID, msg[1]);
-		break;
-	case 0b1110'0000:
-		fluid_synth_pitch_bend(synth.get(), chanID, msg[1] + (msg[2] << 7));
-		break;
-	default: {
-		uint64_t tmp;
-		memcpy(&tmp, msg, sizeof(tmp));
-		LOG_MSG("FSYNTH: unknown MIDI command: %0" PRIx64, tmp);
-		break;
+	if (channel->WakeUp()) {
+		last_rendered_ms = now_ms;
+		return 0;
 	}
+	if (last_rendered_ms >= now_ms) {
+		return 0;
 	}
+	// Return the number of frames needed to get current again
+	assert(ms_per_frame > 0.0);
+	const auto elapsed_ms = now_ms - last_rendered_ms;
+	const auto frames     = iround(ceil(elapsed_ms / ms_per_frame));
+	last_rendered_ms += (frames * ms_per_frame);
+	return check_cast<uint16_t>(frames);
 }
 
-void MidiHandlerFluidsynth::PlaySysex(uint8_t *sysex, size_t len)
+// The request to play the channel message is placed in the MIDI work FIFO
+void MidiHandlerFluidsynth::PlayMsg(const uint8_t* msg)
 {
-	assert(channel);
-	channel->WakeUp();
+	constexpr auto channel_message_len = 4;
+	std::vector<uint8_t> message(msg, msg + channel_message_len);
+	MidiWork work{std::move(message), GetNumPendingFrames(), MessageType::Channel};
+	work_fifo.Enqueue(std::move(work));
+}
 
-	const char *data = reinterpret_cast<const char *>(sysex);
-	const auto n = static_cast<int>(len);
-	fluid_synth_sysex(synth.get(), data, n, nullptr, nullptr, nullptr, false);
+// The request to play the sysex message is placed in the MIDI work FIFO
+void MidiHandlerFluidsynth::PlaySysex(uint8_t* sysex, size_t len)
+{
+	std::vector<uint8_t> message(sysex, sysex + len);
+	MidiWork work{std::move(message), GetNumPendingFrames(), MessageType::Sysex};
+	work_fifo.Enqueue(std::move(work));
```

Comment:
Same comments here: I'm confused by the `std::vector<uint8_t> message(sysex, sysex + len);` syntax.

Reply:
(mentioned in prior comment)